### PR TITLE
[istio] Delete orphan istio resources after disable module

### PR DESCRIPTION
**Pull request by AndreyPavlovFlant**
## Description

Add a hook to check for orphaned resources and remove them when the module is shut down

## Why do we need it, and what problem does it solve?

If an operator's pod fails to start or was deleted before its resources, they remain in the cluster as orphaned and cannot be deleted because of `finalizers`.
To avoid this, we add a check for orphaned resources and delete them when the module is disabled.

Fixes https://github.com/deckhouse/deckhouse/issues/5285.

## What is the expected result?

After disabling the module, there are no remaining Istio system-components in the cluster.

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: istio
type: fix
summary: After disabling the module, clean up any orphaned Istio components.
impact_level: default
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/aws'...

File: modules/110-istio/hooks/purge_orphan_resources.go [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+			ApiVersion:                   "v1",
+			Kind:                         "ServiceAccount",
+			ExecuteHookOnEvents:          pointer.Bool(false),
+			ExecuteHookOnSynchronization: pointer.Bool(false),
+			FilterFunc:                   applyRemoveObjectFilter,
+			NamespaceSelector: &types.NamespaceSelector{
+				NameSelector: &types.NameSelector{
+					MatchNames: []string{istioSystemNs},
+				},
+			},
```
Comment (apolovov) [inline modules/110-istio/hooks/purge_orphan_resources.go:127]:
```suggestion
			NamespaceSelector: lib.NsSelector()
```

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/aws'...

File: modules/110-istio/hooks/purge_orphan_resources.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+		iops, err := k8sClient.Dynamic().Resource(iopGVR).Namespace(istioSystemNs).List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			return err
+		}
+		for _, iop := range iops.Items {
+			_, err = k8sClient.Dynamic().Resource(iopGVR).Namespace(istioSystemNs).Patch(context.TODO(), iop.GetName(), types.MergePatchType, patch, metav1.PatchOptions{})
+			if err != nil {
+				return err
+			}
+			input.LogEntry.Infof("Remove finalizers from IstioOperator/%s in namespace %s", iop.GetName(), istioSystemNs)
```
Comment (apolovov) [inline modules/110-istio/hooks/purge_orphan_resources.go:77]:
```suggestion
			input.LogEntry.Infof("Finalizers from IstioOperator/%s in namespace %s removed", iop.GetName(), istioSystemNs)
```

File: modules/110-istio/hooks/purge_orphan_resources.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+			_, iopDeletionTimestampExists := iop.GetAnnotations()["deletionTimestamp"]
+			if !iopDeletionTimestampExists {
+				err := k8sClient.Dynamic().Resource(iopGVR).Namespace(istioSystemNs).Delete(context.TODO(), iop.GetName(), metav1.DeleteOptions{})
+				if err != nil {
+					return err
+				}
+				input.LogEntry.Infof("Delete IstioOperator/%s in namespace %s", iop.GetName(), istioSystemNs)
```
Comment (apolovov) [inline modules/110-istio/hooks/purge_orphan_resources.go:84]:
```suggestion
				input.LogEntry.Infof("IstioOperator/%s deleted from namespace %s", iop.GetName(), istioSystemNs)
```

File: modules/110-istio/hooks/purge_orphan_resources.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+		}
+		// }
+		// delete NS
+		_, nsDeletionTimestampExists := ns.GetAnnotations()["deletionTimestamp"]
+		if !nsDeletionTimestampExists {
+			err := k8sClient.CoreV1().Namespaces().Delete(context.TODO(), ns.GetName(), metav1.DeleteOptions{})
+			if err != nil {
+				return err
+			}
+			input.LogEntry.Infof("Delete namespace %s", ns.GetName())
```
Comment (apolovov) [inline modules/110-istio/hooks/purge_orphan_resources.go:95]:
```suggestion
			input.LogEntry.Infof("Namespace %s deleted", ns.GetName())
```

File: modules/110-istio/hooks/purge_orphan_resources.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+			_, iopDeletionTimestampExists := iop.GetAnnotations()["deletionTimestamp"]
+			if !iopDeletionTimestampExists {
+				err := k8sClient.Dynamic().Resource(iopGVR).Namespace(istioSystemNs).Delete(context.TODO(), iop.GetName(), metav1.DeleteOptions{})
+				if err != nil {
+					return err
+				}
+				input.LogEntry.Infof("Delete IstioOperator/%s in namespace %s", iop.GetName(), istioSystemNs)
+			}
+		}
+		// }
```
Comment (apolovov) [inline modules/110-istio/hooks/purge_orphan_resources.go:87]:
```suggestion

```

File: modules/110-istio/hooks/purge_orphan_resources.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+	icrs, err := k8sClient.RbacV1().ClusterRoles().List(context.TODO(), metav1.ListOptions{LabelSelector: istioComponentsLabelSelector})
+	if err != nil {
+		return err
+	}
+	for _, icr := range icrs.Items {
+		err := k8sClient.RbacV1().ClusterRoles().Delete(context.TODO(), icr.GetName(), metav1.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+		input.LogEntry.Infof("Delete ClusterRole/%s", icr.GetName())
```
Comment (apolovov) [inline modules/110-istio/hooks/purge_orphan_resources.go:109]:
```suggestion
		input.LogEntry.Infof("ClusterRole/%s deleted", icr.GetName())
```

File: modules/110-istio/hooks/purge_orphan_resources.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+	icrbs, err := k8sClient.RbacV1().ClusterRoleBindings().List(context.TODO(), metav1.ListOptions{LabelSelector: istioComponentsLabelSelector})
+	if err != nil {
+		return err
+	}
+	for _, icrb := range icrbs.Items {
+		err := k8sClient.RbacV1().ClusterRoleBindings().Delete(context.TODO(), icrb.GetName(), metav1.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+		input.LogEntry.Infof("Delete ClusterRoleBinding/%s", icrb.GetName())
```
Comment (apolovov) [inline modules/110-istio/hooks/purge_orphan_resources.go:122]:
```suggestion
		input.LogEntry.Infof("ClusterRoleBinding/%s deleted", icrb.GetName())
```

File: modules/110-istio/hooks/purge_orphan_resources.go [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+	imwcs, err := k8sClient.AdmissionregistrationV1().MutatingWebhookConfigurations().List(context.TODO(), metav1.ListOptions{LabelSelector: istioComponentsLabelSelector})
+	if err != nil {
+		return err
+	}
+	for _, imwc := range imwcs.Items {
+		err := k8sClient.AdmissionregistrationV1().MutatingWebhookConfigurations().Delete(context.TODO(), imwc.GetName(), metav1.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+		input.LogEntry.Infof("Delete MutatingWebhookConfiguration/%s", imwc.GetName())
```
Comment (apolovov) [inline modules/110-istio/hooks/purge_orphan_resources.go:135]:
```suggestion
		input.LogEntry.Infof("MutatingWebhookConfiguration/%s deleted", imwc.GetName())
```

File: modules/110-istio/hooks/purge_orphan_resources.go [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+	ivwcs, err := k8sClient.AdmissionregistrationV1().ValidatingWebhookConfigurations().List(context.TODO(), metav1.ListOptions{LabelSelector: istioComponentsLabelSelector})
+	if err != nil {
+		return err
+	}
+	for _, ivwc := range ivwcs.Items {
+		err := k8sClient.AdmissionregistrationV1().ValidatingWebhookConfigurations().Delete(context.TODO(), ivwc.GetName(), metav1.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+		input.LogEntry.Infof("Delete ValidatingWebhookConfiguration/%s", ivwc.GetName())
```
Comment (apolovov) [inline modules/110-istio/hooks/purge_orphan_resources.go:148]:
```suggestion
		input.LogEntry.Infof("ValidatingWebhookConfiguration/%s deleted", ivwc.GetName())
```

File: modules/110-istio/hooks/purge_orphan_resources_test.go [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+	"sigs.k8s.io/yaml"
+
+	. "github.com/deckhouse/deckhouse/testing/hooks"
+)
+
+var _ = Describe("Istio hooks :: purge_orphan_resources ::", func() {
+
+	const (
+		istioSystemNs = "d8-istio"
+		// istioComponentsLabelSelector = "install.operator.istio.io/owning-resource-namespace=d8-istio"
```
Comment (apolovov) [inline modules/110-istio/hooks/purge_orphan_resources_test.go:38]:
```suggestion
```

File: modules/110-istio/hooks/purge_orphan_resources_test.go [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+		})
+
+		It("Hook must execute successfully", func() {
+			Expect(f).To(ExecuteSuccessfully())
+			Expect(string(f.LogrusOutput.Contents())).ToNot(HaveLen(0))
+			Expect(f.KubernetesGlobalResource("Namespace", "d8-istio").Exists()).To(BeFalse())
+			Expect(string(f.LogrusOutput.Contents())).To(ContainSubstring("Delete namespace d8-istio"))
+			Expect(f.KubernetesResource("IstioOperator", "d8-istio", "v1x16").Exists()).To(BeFalse())
+			Expect(string(f.LogrusOutput.Contents())).To(ContainSubstring("Remove finalizers from IstioOperator/v1x16 in namespace d8-istio"))
+			Expect(string(f.LogrusOutput.Contents())).To(ContainSubstring("Delete IstioOperator/v1x16 in namespace d8-istio"))
```
Comment (apolovov) [inline modules/110-istio/hooks/purge_orphan_resources_test.go:177]:
```suggestion
			Expect(string(f.LogrusOutput.Contents())).To(ContainSubstring("Finalizers from IstioOperator/v1x16 in namespace d8-istio removed"))
			Expect(string(f.LogrusOutput.Contents())).To(ContainSubstring("IstioOperator/v1x16 deleted from namespace d8-istio"))
```

File: modules/110-istio/hooks/purge_orphan_resources_test.go [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+			Expect(string(f.LogrusOutput.Contents())).To(ContainSubstring("Delete ClusterRole/istio-reader-clusterrole-v1x16-d8-istio"))
+			Expect(f.KubernetesGlobalResource("ClusterRoleBinding", "istio-reader-d8-istio").Exists()).To(BeFalse())
+			Expect(string(f.LogrusOutput.Contents())).To(ContainSubstring("Delete ClusterRoleBinding/istio-reader-d8-istio"))
+			Expect(f.KubernetesGlobalResource("ClusterRoleBinding", "istio-reader-clusterrole-v1x16-d8-istio").Exists()).To(BeFalse())
+			Expect(string(f.LogrusOutput.Contents())).To(ContainSubstring("Delete ClusterRoleBinding/istio-reader-clusterrole-v1x16-d8-istio"))
+			Expect(f.KubernetesGlobalResource("MutatingWebhookConfiguration", "istio-sidecar-injector-v1x16-d8-istio").Exists()).To(BeFalse())
+			Expect(string(f.LogrusOutput.Contents())).To(ContainSubstring("Delete MutatingWebhookConfiguration/istio-sidecar-injector-v1x16-d8-istio"))
+			Expect(f.KubernetesGlobalResource("ValidatingWebhookConfiguration", "istio-validator-v1x16-d8-istio").Exists()).To(BeFalse())
+			Expect(string(f.LogrusOutput.Contents())).To(ContainSubstring("Delete ValidatingWebhookConfiguration/istio-validator-v1x16-d8-istio"))
```
Comment (apolovov) [inline modules/110-istio/hooks/purge_orphan_resources_test.go:219]:
```suggestion
			Expect(string(f.LogrusOutput.Contents())).To(ContainSubstring("Namespace d8-istio deleted"))

			Expect(f.KubernetesResource("IstioOperator", "d8-istio", "v1x16").Exists()).To(BeFalse())
			Expect(string(f.LogrusOutput.Contents())).To(ContainSubstring("Finalizers removed from IstioOperator/v1x16 in namespace d8-istio"))
			Expect(string(f.LogrusOutput.Contents())).To(ContainSubstring("IstioOperator/v1x16 deleted from namespace d8-istio"))

			Expect(f.KubernetesGlobalResource("ClusterRole", "istio-reader-d8-istio").Exists()).To(BeFalse())
			Expect(string(f.LogrusOutput.Contents())).To(ContainSubstring("ClusterRole/istio-reader-d8-istio deleted"))
			Expect(f.KubernetesGlobalResource("ClusterRole", "istio-reader-clusterrole-v1x16-d8-istio").Exists()).To(BeFalse())
			Expect(string(f.LogrusOutput.Contents())).To(ContainSubstring("ClusterRole/istio-reader-clusterrole-v1x16-d8-istio deleted"))
			Expect(f.KubernetesGlobalResource("ClusterRoleBinding", "istio-reader-d8-istio").Exists()).To(BeFalse())
			Expect(string(f.LogrusOutput.Contents())).To(ContainSubstring("ClusterRoleBinding/istio-reader-d8-istio deleted"))
			Expect(f.KubernetesGlobalResource("ClusterRoleBinding", "istio-reader-clusterrole-v1x16-d8-istio").Exists()).To(BeFalse())
			Expect(string(f.LogrusOutput.Contents())).To(ContainSubstring("ClusterRoleBinding/istio-reader-clusterrole-v1x16-d8-istio deleted"))
			Expect(f.KubernetesGlobalResource("MutatingWebhookConfiguration", "istio-sidecar-injector-v1x16-d8-istio").Exists()).To(BeFalse())
			Expect(string(f.LogrusOutput.Contents())).To(ContainSubstring("MutatingWebhookConfiguration/istio-sidecar-injector-v1x16-d8-istio deleted"))
			Expect(f.KubernetesGlobalResource("ValidatingWebhookConfiguration", "istio-validator-v1x16-d8-istio").Exists()).To(BeFalse())
			Expect(string(f.LogrusOutput.Contents())).To(ContainSubstring("ValidatingWebhookConfiguration/istio-validator-v1x16-d8-istio deleted"))
```

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/aws'...

**Review by apolovov — APPROVED** [review]

**Pull request merged**